internal/cloud/gcp/cloudbuild.go

```diff
@@ -0,0 +1,224 @@
+//
+// Uses:
+//	- Storage API
+//	- Cloud Build API
+//	- Compute Engine API (indirectly)
+func (g *GCP) CloudbuildBuildCleanup(ctx context.Context, buildID string) ([]string, error) {
+	var deletedResources []string
+
+	storageClient, err := storage.NewClient(ctx, option.WithCredentials(g.creds))
+	if err != nil {
+		return deletedResources, fmt.Errorf("failed to get Storage client: %v", err)
+	}
+	defer storageClient.Close()
+
+	buildLog, err := g.CloudbuildBuildLog(ctx, buildID)
+	if err != nil {
+		return deletedResources, fmt.Errorf("failed to get log for build ID '%s': %v", buildID, err)
+	}
+
+	resources, err := cloudbuildResourcesFromBuildLog(buildLog)
+	if err != nil {
+		return deletedResources, fmt.Errorf("extracting created resources from build log failed: %v", err)
+	}
+
+	// Delete all Compute Engine instances
+	for _, instance := range resources.computeInstances {
+		err = g.ComputeInstanceDelete(ctx, resources.zone, instance)
+		if err == nil {
+			deletedResources = append(deletedResources, fmt.Sprintf("instance: %s (%s)", instance, resources.zone))
+		}
+	}
+
+	// Deleting instances in reality takes some time. Deleting a disk while it is still used by the instance, will fail.
+	// Iterate over the list of instances and wait until they are all deleted.
+	for _, instance := range resources.computeInstances {
+		for {
+			_, err := g.ComputeInstanceGet(ctx, resources.zone, instance)
+			// Getting the instance information failed, it is ideleted.
+			if err != nil {
+				break
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}
+
+	// Delete all Compute Engine Disks
+	// Deleting Compute instances takes time, therefore it may happen that deleting the disk will fail, because
+	// it is still in use by the instance. Retry in such case
+	for _, disk := range resources.computeDisks {
+		err = g.ComputeDiskDelete(ctx, resources.zone, disk)
+		if err == nil {
+			deletedResources = append(deletedResources, fmt.Sprintf("disk: %s (%s)", disk, resources.zone))
+		}
+	}
+
+	// Delete all Storage cache files
+	bucket := storageClient.Bucket(resources.storageCacheDir.bucket)
+	objects := bucket.Objects(ctx, &storage.Query{Prefix: resources.storageCacheDir.dir})
+	for {
+		objAttrs, err := objects.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			// Do not return, just continue with the next object
+			continue
+		}
+
+		object := storageClient.Bucket(objAttrs.Bucket).Object(objAttrs.Name)
+		if err = object.Delete(ctx); err == nil {
+			deletedResources = append(deletedResources, fmt.Sprintf("storage object: %s/%s", objAttrs.Bucket, objAttrs.Name))
+		}
+	}
+
+	return deletedResources, nil
+}
+
+// cloudbuildResourcesFromBuildLog parses the provided Cloud Build log for any
+// resources that were created by the job as part of its work. The list of extracted
+// resources is returned as cloudbuildBuildResources struct instance
+func cloudbuildResourcesFromBuildLog(buildLog string) (*cloudbuildBuildResources, error) {
```

Comment:
Wow, there are no functions for this in the official GCP API?

Reply:
No, more info in https://issuetracker.google.com/issues/182680680